### PR TITLE
x/gov: allow to change rule quorum

### LINKED ISSUE #869: Change an election quorum
Ability to create a proposal by governing board to change an election quorum.

## Acceptance criteria
- [x] Modify `UpdateElectionRuleMsg` in the `gov` package to allow to change an election quorum.
- [x] Update `bnscli` to include creation of this message.

### PR DESCRIPTION
**Pull request by husio**
Update `UpdateElectionRuleMsg` and add `quorum` attribute to allow
overwriting existing election configuration.

`bnscli` updated to allow for the attribute setup.

resolve #869

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: cmd/bnscli/cmd_gov.go [hunk 1]
```diff
@@ -485,6 +485,7 @@ Creates a new version for an existing election rule. The new version is used for
 		durationFl    = fl.Int("voting-period", 0, "Duration in seconds how long the voting period will take place")
 		numeratorFl   = fl.Int("threshold-numerator", 0, "The top number of the fraction.")
 		denominatorFl = fl.Uint("threshold-denominator", 0, "The bottom number of the fraction")
+		quorumFl      = flFraction(fl, "quorum", "0/1", "New quorum fraction in format <numerator>/<denominator>. Zero quorum deletes the value.")
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_gov.go]: I am skeptical of the default.
If I don't provide a value, it should be a no-op, not changing the quroum

File: cmd/bnscli/cmd_gov.go [hunk 2]
```diff
@@ -498,13 +499,20 @@ Creates a new version for an existing election rule. The new version is used for
 	if err := fraction.Validate(); err != nil {
 		flagDie("invalid voting period: %s", err)
 	}
+
+	var quorum *gov.Fraction
+	if f := quorumFl.Fraction(); f.Numerator != 0 {
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_gov.go]: Ah... okay, if you set to 0, it ignore this value (not deletes it).
Can you improve the comment?

File: cmd/bnscli/flag_test.go [hunk 3]
```diff
@@ -302,3 +303,77 @@ func observeFlagDie(t testing.TB) (*int, func()) {
 	}
 	return &cnt, cleanup
 }
+
+func TestFractionFlag(t *testing.T) {
```
Comment (ethanfrey) [inline cmd/bnscli/flag_test.go]: So much code just for fraction support.

File: x/gov/handler.go [hunk 4]
```diff
@@ -542,6 +542,7 @@ func (h UpdateElectionRuleHandler) Deliver(ctx weave.Context, db weave.KVStore,
 	}
 	rule.Threshold = msg.Threshold
 	rule.VotingPeriod = msg.VotingPeriod
+	rule.Quorum = msg.Quorum
```
Comment (ethanfrey) [inline x/gov/handler.go]: Only set this if not-nil.

Consider behavior.
I have a rule with 1/2 threshold and 3/5 quorum.
I want to update it to 3/5 threshold. If I do not mention the quorum, it gets removed?

Or maybe this is right... I keep thinking of PATCH actions and you keep writing POST logic. Maybe I just need to change my verb

**Comment by husio** [comment]
I have improved the implementation and now the quorum value is set only if provided. This is [in the `bnscli`](
https://github.com/iov-one/weave/pull/886/files#diff-3389fea13f140b1d6ea101d41d03d982R504) only.

As you noticed, the value is always [overwritten in the handler](https://github.com/iov-one/weave/pull/886/files#diff-c51941a9f6c50cb3936120ad31fa8b98R587) regardless the value. This is the same logic that is implemented for the [existing fields](https://github.com/iov-one/weave/blob/master/x/gov/handler.go#L585-L586).

I can change the documentation and make this field mandatory in `bnscli` so that you always have to provide a value (no default).
I can also change the handler implementation instead. Assign value only if it is not `nil`. You have to use a zero denominator to delete. This breaks compatibility as I would have to change how other attributes are set for consistency.

**Pull request merged**